jena-arq/src/main/java/org/apache/jena/sparql/core/Transactional.java

```diff
@@ -192,4 +174,110 @@ public default boolean promote() {
 
     /** Say whether inside a transaction. */
     public boolean isInTransaction() ;
+
+    /**
+     * Execute and return a value in a transaction with the given {@link TxnType transaction type}.
+     */
+    public default <T extends Transactional, X> X calc(TxnType txnType, Supplier<X> r) {
```

Comment:
The downside of `default` methods is that since they become part of the interface implementations could choose to override them especially since you cannot declare them to be `final`

The advantage of the existing `Txn` helper class is that it provides a predictable and consistent behaviour regardless of the `Transactional` implementation it is operating over i.e. if you are using `Txn.executeRead()` it's always going to behave the same whereas calling `ds.executeRead()` could change behaviour depending on the dataset implementation.